File: templates/java/src/main/java/de/finatix/StandoffGameArgs.java
package de.finatix;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.Arrays;

public record StandoffGameArgs(
        PlayerState me,
        PlayerState neighborLeft,
        PlayerState neighborRight
) {
    record PlayerState(int hp, int ammo, Action lastAction) {
        public static PlayerState fromString(String raw) {
            String[] attributes = raw.split("/");
            if (attributes.length != 3) {
                throw new IllegalArgumentException("Expected 3 attributes, got %s".formatted(attributes.length));
            }
            return new PlayerState(Integer.parseInt(attributes[0]), Integer.parseInt(attributes[1]), Action.fromString(attributes[2]));
        }
    }

    @Getter
    @AllArgsConstructor
    enum Action {
        NOTHING("N"),
        SHOOT("S"),
        PROTECT("P"),
        RELOAD("R"),
        ;

        private final String encoded;

        public static Action fromString(String raw) {
            return Arrays.stream(Action.values())
                    .filter(a -> raw.equals(a.encoded))
                    .findFirst()
                    .orElseThrow(() -> new IllegalArgumentException("Unknown action: %s".formatted(raw)));
        }
    }
}

File: templates/python/src/standoff_game_args.py
from __future__ import annotations
from dataclasses import dataclass
from enum import StrEnum


class Action(StrEnum):
    SHOOT = "S"
    PROTECT = "P"
    RELOAD = "R"
    NOTHING = "N"

    @staticmethod
    def from_str(s: str) -> Action:
        if s == "S":
            return Action.SHOOT
        elif s == "P":
            return Action.PROTECT
        elif s == "R":
            return Action.RELOAD
        elif s == "N":
            return Action.NOTHING
        else:
            raise Exception(f"Unknown action: {s}")


@dataclass(frozen=True)
class PlayerState:
    hp: int
    ammo: int
    last_action: Action

    @staticmethod
    def from_str(raw: str) -> PlayerState:
        attr = raw.split("/")
        if len(attr) != 3:
            raise Exception(f"Expected 3 arguments, got {len(attr)}")
        return PlayerState(int(attr[0]), int(attr[1]), Action.from_str(attr[2]))


@dataclass(frozen=True)
class StandoffGameArgs:
    me: PlayerState
    neighbor_left: PlayerState
    neighbor_right: PlayerState
